### PR TITLE
6345 store user data size in redis

### LINKED ISSUE #6345: Store UserDataSize in Redis
We need Hubspot to have the actual disk space a user is using.

- [X] Store `UserDataSize` (the value returned by CDB_UserDataSize) in Redis when user enters the dashboard.

cc @vuxita

### PR DESCRIPTION
**Pull request by juanignaciosl**
This closes #6345.

@rafatower could you take a look, please? This might not be the final version of the code, since Hubspot reporting side is not done yet but, since you helped a lot with the design, I'd like your feedback.

I considered `ZADD`, but I think sorting by score (`db_size_in_bytes` in my case) is not useful. In my original proposal (and code) I added the date to the key (`rails:users:central-user-1:db_size_in_bytes:20160105`) but it was not actually needed, so I removed it to avoid duplications (`rails:users:central-user-1:db_size_in_bytes`).

I had also recovered the original resque job for this. Originally the code had to do an http call to a third party service, so a job made sense, but now we're only storing a Redis value, the job is an unnecessary overhead and I've removed it (you can see the latest commits).

What do you think?

### CODE REVIEW

**Comment by juanignaciosl** [comment]
@rafatower hold this review. Currently there's no way for Central to read cloud Redis, so this design must be changed :(

**Comment by juanignaciosl** [comment]
FYI, since CartoDB Redis is not a valid communication channel with Central, I'm going to modify superadmin user api so it can be filtered by "disk_space_change", in a similar fashion than "overquota" filtering.

**Comment by juanignaciosl** [comment]
@rafatower , @zenitraM , could you CR this, please?

File: lib/carto/users_metadata_redis_cache.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        set_db_size_in_bytes(user)
+      end
+    end
+
+    def db_size_in_bytes(user)
+      @redis.get(db_size_in_bytes_key(user.username)).to_i
+    end
+
+    def db_size_in_bytes_change_users
+      keys = @redis.scan_each(match: db_size_in_bytes_key('*')).to_a.uniq
```
Comment (rafatower) [inline lib/carto/users_metadata_redis_cache.rb]: This is iterating on `rails:users:*:db_size_in_bytes`. Wondering if that could be a problem as we add more stuff into `rails:users:...`

File: lib/carto/users_metadata_redis_cache.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
+# encoding: utf-8
+
+module Carto
+  class UsersMetadataRedisCache
```
Comment (rafatower) [inline lib/carto/users_metadata_redis_cache.rb]: As it is only used for db_size, would you consider other names? maybe `UserDbSizeCache`

File: lib/carto/users_metadata_redis_cache.rb (hunk 1 above)
Comment (juanignaciosl) [inline lib/carto/users_metadata_redis_cache.rb]: Not as long as new keys don't finish with `db_size_in_bytes`, isn't it?

What do you think about using existing Redis `$users_metadata` database (number 5) for this? // @zenitraM

File: app/models/user.rb [hunk 3]
```diff
@@ -456,6 +456,10 @@ def self.overquota(delta = 0)
     end
   end
 
+  def self.db_size_in_bytes_change
```
Comment (rafatower) [inline app/models/user.rb]: this is returning a collection of users whose db's changed in size, right? I'd add a comment and/or rename to `where_db_size_changed` for clarity

File: lib/carto/users_metadata_redis_cache.rb (hunk 2 above)
Comment (juanignaciosl) [inline lib/carto/users_metadata_redis_cache.rb]: Maybe I over-engineered this. I was thinking about encapsulating all user metadata cache in this class, that's why `update_if_old` receives `user`, not `db_size_in_bytes`. I'd prefer keeping this class name and `update` signature, but certainly this is open to debate. Do you prefer renaming both the class (`UserDbSizeCache`) and the method (`update_db_size_if_old`), or keeping both as they are?

File: lib/carto/users_metadata_redis_cache.rb (hunk 1 above)
Comment (rafatower) [inline lib/carto/users_metadata_redis_cache.rb]: $users_metadata is ok, I wouldn't consider another db

File: lib/carto/users_metadata_redis_cache.rb (hunk 2 above)
Comment (rafatower) [inline lib/carto/users_metadata_redis_cache.rb]: one of them is enough I think

File: lib/carto/users_metadata_redis_cache.rb (hunk 1 above)
Comment (zenitraM) [inline lib/carto/users_metadata_redis_cache.rb]: +1 to rafa, also be sure this uses a SCAN behind the scenes on Redis and not a KEYS, KEYS is evil but I think SCAN is good.

Comment (juanignaciosl) [inline lib/carto/users_metadata_redis_cache.rb]: It does :-)

```
  def scan_each(options={}, &block)
    return to_enum(:scan_each, options) unless block_given?
    cursor = 0
    loop do
      cursor, keys = scan(cursor, options)
      keys.each(&block)
      break if cursor == "0"
    end
  end
```

```
  def scan(cursor, options={})
    _scan(:scan, cursor, [], options)
  end
```

There's no better approach for this than `SCAN` + `MGET`, isn't it?

File: app/models/user.rb (hunk 3 above)
Comment (juanignaciosl) [inline app/models/user.rb]: True, good catch. Renamed to `self.where_db_size_in_bytes_change`, and removed `.all` from the method so it doesn't return a list of users but a query you can keep filtering.

File: lib/carto/users_metadata_redis_cache.rb (hunk 2 above)
Comment (juanignaciosl) [inline lib/carto/users_metadata_redis_cache.rb]: Renamed to `UserDbSizeCache` :+1:

**Comment by juanignaciosl** [comment]
Retest this, please.

**Comment by juanignaciosl** [comment]
@rafatower , a final CR after your suggestions, please? @zenitraM , I'd also like you to CR this, since it involves Redis. Thank you both :-)

**Comment by zenitraM** [comment]
:+1: for me

File: spec/requests/superadmin/users_spec.rb [hunk 4]
```diff
@@ -1,7 +1,12 @@
 # encoding: utf-8
+
+require 'ostruct'
 require_relative '../../acceptance_helper'
 
 feature "Superadmin's users API" do
+  before(:all) do
+  end
```
Comment (rafatower) [inline spec/requests/superadmin/users_spec.rb]: minor comment: can be deleted?

**Comment by rafatower** [comment]
:+1:

File: spec/requests/superadmin/users_spec.rb (hunk 4 above)
Comment (juanignaciosl) [inline spec/requests/superadmin/users_spec.rb]: Sure, I overlooked this, thanks!

**Comment by juanignaciosl** [comment]
I'll merge & deploy after empty code deletion (60f28f36d48e0832beb0d5d7f76d860ce8cb1120) test pass, thank you both!

**Pull request merged**